mesh/meshdb.go

```diff
@@ -145,109 +159,188 @@ func (m *DB) Close() {
 
 // todo: for now, these methods are used to export dbs to sync, think about merging the two packages
 
+// Ballots exports the ballot database.
+func (m *DB) Ballots() database.Getter {
+	return newBallotFetcherDB(m)
+}
+
 // Blocks exports the block database.
 func (m *DB) Blocks() database.Getter {
-	m.blockMutex.RLock()
-	defer m.blockMutex.RUnlock()
-	return NewBlockFetcherDB(m)
+	return newBlockFetcherDB(m)
 }
 
 // Transactions exports the transactions DB.
 func (m *DB) Transactions() database.Getter {
 	return m.transactions
 }
 
-// ErrAlreadyExist error returned when adding an existing value to the database.
-var ErrAlreadyExist = errors.New("block already exists in database")
-
 // HasBallot returns true if the database has Ballot specified by the BallotID and false otherwise.
-func (m *DB) HasBallot(types.BallotID) bool {
-	// TODO: implement me
-	return false
+func (m *DB) HasBallot(id types.BallotID) bool {
+	has, err := m.ballots.Has(id.Bytes())
+	return err == nil && has
+}
+
+// AddBallot adds a ballot to the database.
+func (m *DB) AddBallot(b *types.Ballot) error {
+	m.blockMutex.Lock()
+	defer m.blockMutex.Unlock()
+	if m.HasBallot(b.ID()) {
+		m.With().Warning("ballot already exists", b.ID())
+		return nil
+	}
+	if err := m.writeBallot(b); err != nil {
+		return fmt.Errorf("write ballot: %w", err)
+	}
+	return nil
 }
 
 // GetBallot returns true if the database has Ballot specified by the BallotID and false otherwise.
-func (m *DB) GetBallot(ballotID types.BallotID) (*types.Ballot, error) {
-	block, err := m.GetBlock(types.BlockID(ballotID))
-	if err == nil {
-		return block.ToBallot(), nil
+func (m *DB) GetBallot(id types.BallotID) (*types.Ballot, error) {
+	data, err := m.getBallotBytes(id)
+	if err != nil {
+		return nil, err
+	}
+	b := &types.Ballot{}
+	if err := types.BytesToInterface(data, b); err != nil {
+		return nil, fmt.Errorf("parse ballot: %w", err)
 	}
-	return nil, err
+	if err := b.Initialize(); err != nil {
```

Comment:
should not be called when loading from db

Reply:
I think we should be smart about when to do this costly initialization.

When receiving a ballot over the wire, we obviously have to calculate the ballot ID and extract the public key from the signature for validation.

When fetching from the DB, since the key includes the ID we should inject that into the object before returning it.

The author smesher ID should be calculated lazily and then cached. In most cases I don't expect to need it, but if we do need it in a performance critical code path, we should store it in the DB.

Reply:
changed to such that SmesherID() is calculated eagerly upon receiving the data, but lazily when retrieved from DB.

Reply:
Didn't look at the code, but for simplicity it can always be lazy. When receiving we will immediately try to validate the ballot, so this will trigger the ID calculation.

Reply:
changed the code as @dshulyak suggested (restored the previous behavior).